Allow specifying tags for aten operators in native_functions.yaml

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72549

This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

anjali411 committed:
Allow specifying tags for aten operators in native_functions.yaml

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -730,7 +730,8 @@
   device_guard: False
-  tags: inplace_view
+  tags:
+    is_inplace_view: True
```
This format seems... worse? Do we have use cases for non-boolean valued tags?

ezyang commented:
oops, didn't see not for review. Unsubscribing for now, please re-cc me when ready

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


Not ready for review

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"




[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"




[ghstack-poisoned]

ezyang commented on tools/autograd/gen_autograd.py:
```diff
@@ -41,28 +41,30 @@
     differentiability_infos = load_derivatives(
-        os.path.join(autograd_dir, 'derivatives.yaml'), native_functions_path)
+        os.path.join(autograd_dir, 'derivatives.yaml'), native_functions_path,
+        tags_path)
```
this probably fails flake

ezyang commented:
These tags still aren't exposed from Python, is that right? It would be nice to parse in the descriptions so that we can then generate appropriate `__doc__` on the Python side representation (seems better than slinging strings around in Python).

ezyang commented on tools/codegen/gen.py:
```diff
@@ -144,12 +145,38 @@ def parse_native_yaml_struct(es: object, path: str = "<stdin>") -> ParsedYaml:
+    # TODO: parse tags.yaml and create a tags database (a dict of tag name mapping to a Tag object)
+    with open(path, 'r') as f:
+        es = yaml.load(f, Loader=LineLoader)
+        ValidTags = parse_tags_yaml_struct(es, path=path)
```
nit: upper-cased identifier here isn't very idiomatic

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 commented:
> ValidTags

no they aren't yet but the next PR in python

> These tags still aren't exposed from Python, is that right? It would be nice to parse in the descriptions so that we can then generate appropriate `__doc__` on the Python side representation (seems better than slinging strings around in Python).

No they aren't yet but the next PR in the stack exposes the tags through the torch.ops API.

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

malfet commented on BUILD.bazel:
```diff
@@ -185,6 +185,7 @@ genrule(
     name = "all_generated_code",
     srcs = [
         "aten/src/ATen/native/native_functions.yaml",
+        "aten/src/ATen/native/tags.yaml"
```
I guess that was the source of the problem :)
```suggestion
        "aten/src/ATen/native/tags.yaml",
```

malfet committed:
Update BUILD.bazel

anjali411 commented:
@pytorchbot merge this please

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 commented:
@pytorchbot merge this please

anjali411 committed:
Update on "Allow specifying tags for aten operators in native_functions.yaml"


This PR:
1. Adds a new yaml file. For each new tag, we will require an entry in this file along with a mandatory description.
2. There are checks to ensure that any tags added for native_functions are valid tags (i.e., they exist in tags.yaml)
3. Each NativeFunction object holds tags as a set of strings (this logic can be found in model.py)

Most file changes are to update the call to `parse_native_yaml` to include the path to `native_functions.yaml` as well as `tags.yaml`

[ghstack-poisoned]

anjali411 commented:
@pytorchmergebot merge this please

malfet commented:
@pytorchbot revert this

malfet commented:
Reverting as this change broke internal build systems. Significant build system changes still have to go thru the old land path
